### ISSUE TITLE
Question: What is the size in bytes of a null pointer?

### ISSUE THREAD

**Issue by jtlz2** [pleiszenburg/zugbruecke#35]
In `rpc.py`, what should ` size_bytes` be when the variable is a void pointer, i.e.

C Prototype:

```
typedef struct
{
        void             *F;
}A;
```

ctypes/ZB Struct:

```
class A(ctypes.Structure):
    _fields_ = [("F", ctypes.c_void_p)]

func.memsync = [
        {
                'p': [0,'F'],
                'l': [WHAT TO GIVE HERE?],
                '_t': [WHAT TO GIVE HERE?]
        }
        ]
```

```
def serialize_pointer_into_int_list(ctypes_pointer, size_bytes):

        return (ctypes.c_ubyte * size_bytes).from_address(ctypes.c_void_p.from_buffer(
                ctypes.cast(ctypes_pointer, ctypes.POINTER(ctypes.c_ubyte * size_bytes))
                ).value)[:]
```

At the moment that throws:

```
Traceback (most recent call last):
  File "./func.py", line 1125, in <module>
    ret=main()
  File "./func.py", line 949, in main
    status=func(a,b,c)
  File "/usr/local/lib/python3.6/site-packages/zugbruecke/core/routine_client.py", line 116, in __call__
    mem_package_list, memory_transport_handle = self.data.client_pack_memory_list(args, self.memsync)
  File "/usr/local/lib/python3.6/site-packages/zugbruecke/core/data/memory.py", line 106, in client_pack_memory_list
    data = serialize_pointer_into_int_list(arg_value, length_value)
  File "/usr/local/lib/python3.6/site-packages/zugbruecke/core/memory.py", line 53, in serialize_pointer_into_int_list
    ctypes.cast(ctypes_pointer, ctypes.POINTER(ctypes.c_ubyte * size_bytes))
TypeError: integer expected
```

Thanks!

**Comment by s-m-e** [comment]
@jtlz2 Can you provide any context about what ``F`` is? Is it a function, data, something you actually do not need?

**Comment by jtlz2** [comment]
@s-m-e Yep it's a function, but see below. The C prototype is:

```
typedef int(*G)(void*, int*);

typedef struct
{
        G       F;
}A;
```

In the C wrapper, `F` can legitimately be set to `NULL` though, which is why I thought a null pointer would be fine rather than a full-on callback:

```
F = NULL;
```

**Comment by s-m-e** [comment]
@jtlz2 Ok, I tried this. I was not entire sure that my code would handle it correctly - apparently, it does, so I just added another test to make sure it keeps handling it correctly :) Have a look at [this test file](https://github.com/pleiszenburg/zugbruecke/blob/36aff03503518f82052fc9e8f5bd087fbd075d1a/tests/test_callback_optional.py#L47). It's calling into [these two identical test functions](https://github.com/pleiszenburg/zugbruecke/blob/36aff03503518f82052fc9e8f5bd087fbd075d1a/demo_dll/demo_dll.c#L395). The trick is to define the not required callback function argument as a ``ctypes.c_void_p`` and (optionally) initialize it with Python's ``None`` (which, at the end of the day, just serves as a placeholder and does not change the pointer, which already is a Null pointer).

**Comment by jtlz2** [comment]
@s-m-e Thank you so so much for looking into this! Since I am not completely sure how to handle the translation from `int(*G)(void*, int*)`, what I have done is almost certainly wrong:

```
g = ctypes.WINFUNCTYPE(ctypes.c_int)
@g
def G():
    """
    typedef int(*G)(void*, int*);
    """
    return

class A(ctypes.Structure):
    _fields_ = [("F", ctypes.POINTER(g))]
```

Then I just set

```
a=A()
a.F=None
```

Where is my opportunity to set the argument of the callback to `ctypes.c_void_p`?

Am I remotely on the right lines or not really? :\

**Comment by jtlz2** [comment]
@s-m-e I also have another void pointer, the C prototype being:

```
typedef struct
{
        void             *W;
}B;
```

And so in python:

```
class B(ctypes.Structure):
    _fields_ = [("W", ctypes.c_void_p)]

```
When I call `status=func(a,b,c)`, do I need to include `W` in `func.memsync`, and if so what would I give for its `'l'` size?

```
        {
                'p': [1,'W'],
                'l': [WHAT TO GIVE HERE?],
                '_t': ctypes.c_void_p
        },
```

**Comment by s-m-e** [comment]
@jtlz2 Your function prototype `int(*G)(void*, int*)` translates as follows, see [`ctypes` manual](https://docs.python.org/3/library/ctypes.html#ctypes.WINFUNCTYPE):
```
g = ctypes.WINFUNCTYPE(ctypes.c_int, ctypes.c_void_p, ctypes.c_int)
```
Your Python function `G` must accept two parameters and return an integer.
```
@g
def G(param_a, param_b):
    """
    typedef int(*G)(void*, int*);
    """
    return some_integer
```
`g` already is a pointer type, so your structure would look somewhat like this:
```
class A(ctypes.Structure):
    _fields_ = [("F", g)]
```
The problem with the above is that `param_a` likely requires a `memsync` directive to become fully operational (if you're actually using it for something). I am implementing this as part of #25. 

If the field `W` in your struct `B` is just an empty placeholder and is not used for anything, you do not need a `memsync` directive for it (for example if it is an unused/optional field for a function pointer). If on the other hand it is supposed to actually carry some data, you need to figure out its length somehow.

**Comment by jtlz2** [comment]
@s-m-e Many thanks! I think this closes the original question off - makes sense. I will tackle `G` when it comes to your fix for #25 :)
